crates/planner/src/logical/bind.rs

```diff
@@ -0,0 +1,414 @@
+use std::sync::Arc;
+
+use spacetimedb_lib::{from_hex_pad, Address, AlgebraicValue, Identity};
+use spacetimedb_sats::{AlgebraicType, ProductType, ProductTypeElement};
+use spacetimedb_schema::schema::TableSchema;
+use spacetimedb_sql_parser::{
+    ast::{
+        self,
+        sub::{SqlAst, SqlSelect},
+        BinOp, ProjectElem, SqlExpr, SqlFrom, SqlLiteral,
+    },
+    parser::sub::parse_subscription,
+};
+
+use super::errors::{ConstraintViolation, ResolutionError, TypingError, Unsupported};
+use super::expr::{Expr, Ref, RelExpr, Type, Vars};
+
+pub type TypingResult<T> = core::result::Result<T, TypingError>;
+
+pub trait SchemaView {
+    fn schema(&self, name: &str, case_sensitive: bool) -> Option<Arc<TableSchema>>;
+}
+
+/// Parse and type check a subscription query
+pub fn parse_and_type_sub(sql: &str, tx: &impl SchemaView) -> TypingResult<RelExpr> {
+    expect_table_type(type_ast(parse_subscription(sql)?, tx)?)
+}
+
+/// Type check and lower a [SqlAst] into a [RelExpr].
+/// This includes name resolution and variable binding.
+pub fn type_ast(expr: SqlAst, tx: &impl SchemaView) -> TypingResult<RelExpr> {
+    match expr {
+        SqlAst::Union(a, b) => {
+            let a = type_ast(*a, tx)?;
+            let b = type_ast(*b, tx)?;
+            assert_eq_types(a.ty(), b.ty())?;
+            Ok(RelExpr::Union(Box::new(a), Box::new(b)))
+        }
```

Comment:
Remember to add a comment or check here about checking the schemas are compatible with the union/minus projection